### PR TITLE
[Github #627] shared refactor groundwork

### PR DESCRIPTION
**Pull request by khaliqgant**
* Adds a `shared` package and moves a lot of components to the that package for easier future reuse across different packages

### CODE REVIEW

File: packages/shared/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
+    "compilerOptions": {
```
Comment (bastienbeurier) [inline packages/shared/tsconfig.json:3]: Shouldn't we align this one with the `packages/server`'s `tsconfig.json`? 

```
{
    "extends": "../../tsconfig.json",
    "compilerOptions": {
        "rootDir": "lib",
        "outDir": "dist"
    },
    "include": ["lib/**/*", "db/**/*"]
}
```

**Comment by bastienbeurier** [comment]
`packages/server/Dockerfile` should probably be changed to to also copy the `packages/shared` folder, at least that's how we did it for the previous Nango Sync.

https://github.com/NangoHQ/nango-sync/blob/main/packages/server/Dockerfile

**Review by bastienbeurier — APPROVED** [review]

File: packages/shared/tsconfig.json (hunk 1 above)
Comment (khaliqgant) [reply packages/shared/tsconfig.json:3]: Yup, updated with https://github.com/NangoHQ/nango/pull/635/commits/9518da483a7b12a8d5544646be39d5907b14efbf

**Comment by khaliqgant** [comment]
> `packages/server/Dockerfile` should probably be changed to to also copy the `packages/shared` folder, at least that's how we did it for the previous Nango Sync.
> 
> https://github.com/NangoHQ/nango-sync/blob/main/packages/server/Dockerfile

Thanks, updated with https://github.com/NangoHQ/nango/pull/635/commits/f5e60700a228af15342eacba9a86eade287208a3

**Review by bastienbeurier — APPROVED** [review]

**Pull request merged**